lib/Basics/Result.h

```diff
@@ -110,16 +133,104 @@ class Result {
     return *this;
   }
 
-  // the default implementation is const, but sub-classes might
-  // really do more work to compute.
-
-  virtual std::string errorMessage() const& { return _errorMessage; }
-  virtual std::string errorMessage() && { return std::move(_errorMessage); }
-
  protected:
   int _errorNumber;
-  std::string _errorMessage;
+  mutable std::unique_ptr<std::string> _errorMessage;
 };
-}
 
+
+template <typename T>
+struct TypedResult {
+  //exception to the rule: "value instead of _value" to allow easier access to members
+  using ValueType = T;
+  ValueType value;
+private:
+  bool _valid = false;
+  Result _result;
+
+public:
+  TypedResult() noexcept(std::is_nothrow_default_constructible<std::string>::value &&
+                         std::is_nothrow_default_constructible<T>::value
+                        )= default;
+
+  template <bool x = std::is_lvalue_reference<T>::value ||
+                     ( !std::is_move_constructible<T>::value &&
+                       !std::is_move_assignable<T>::value )
+           ,typename std::enable_if<x>::type* = nullptr
+           >
+  TypedResult(ValueType value
+             ,int error = TRI_ERROR_NO_ERROR
+             ,std::string message = ""
+             )
+    : value(value)
+    , _valid(true)
+    ,_result(error, std::move(message))
+    {
+      std::cerr << "copy" << std::endl;
+    }
+
+  template <int x = !std::is_lvalue_reference<T>::value
+                    && std::is_move_constructible<T>::value
+           ,typename std::enable_if<x>::type* = nullptr
+           >
+  TypedResult(ValueType value
```

Comment:
why exactly these overloads for `bool`, `int` and `uint32_t`?
Are these just the initial few types we wanted to overload? Do we need to add all other types here?